// docpad-plugin-dateurls/out/dateurls.plugin.js
// Generated by CoffeeScript 1.4.0
var moment, post_date_regex,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

post_date_regex = new RegExp("([0-9]+-)*");

moment = require('moment');

module.exports = function(BasePlugin) {
  var DateUrls;
  return DateUrls = (function(_super) {

    __extends(DateUrls, _super);

    function DateUrls() {
      return DateUrls.__super__.constructor.apply(this, arguments);
    }

    DateUrls.prototype.name = 'dateurls';

    DateUrls.prototype.config = {
      enabled: true,
      documentPath: 'posts',
      cleanurl: false,
      trailingSlashes: false,
      collectionName: null
    };

    DateUrls.prototype.extendCollections = function(opts) {
      var config, docpad;
      config = this.config;
      if (config.collectionName != null) {
        return this;
      }
      docpad = this.docpad;
      config.collectionName = '_dateurlsPosts';
      docpad.setCollection(config.collectionName, docpad.getCollection('documents').findAllLive({
        relativeDirPath: config.documentPath
      }));
      return this;
    };

    DateUrls.prototype.renderBefore = function(opts, next) {
      var collection, config, documents, getFilename, templateData, trailingSlashes;
      collection = opts.collection, templateData = opts.templateData;
      config = this.config;
      if (config.enabled) {
        if (config.cleanurl) {
          getFilename = 'basename';
          trailingSlashes = this.config.trailingSlashes;
        } else {
          getFilename = 'outFilename';
        }
        documents = this.docpad.getCollection(config.collectionName);
        documents.forEach(function(document) {
          var dateUrl;
          dateUrl = moment.utc(document.getMeta('date')).format('/YYYY/MM/DD') + "/" + document.get(getFilename).replace(post_date_regex, '');
          if (config.cleanurl) {
            document.setUrl(dateUrl + (trailingSlashes ? '/' : ''));
            return document.addUrl(dateUrl + (trailingSlashes ? '' : '/'));
          } else {
            return document.setUrl(dateUrl);
          }
        });
      }
      return next();
    };

    return DateUrls;

  })(BasePlugin);
};
